### PR TITLE
New Navigation Model

### PR DESCRIPTION
**Pull request by aintgoin2goa**
Removing the n-navigation middleware and replacing it with a navigationModel that gets it's data from the new API

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -141,13 +143,14 @@ module.exports = function(options) {
...
 			viewsDirectory: options.viewsDirectory
-		});
+		}));
 	}
 
 	// add statutory metadata to construct the page
 	if (options.withNavigation) {
-		flagsPromise.then(navigation.init);
-		app.use(navigation.middleware);
+		let navigation = new NavigationModel({withNavigationHierarchy:options.withNavigationHierarchy});
```
Comment (i-like-robots) [inline main.js]: `const navigation = new NavigationModel(options)`?

File: src/navigation/hierarchyMixin.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	return {item:null, parent:null};
+}
+
+module.exports = class HierarchyMixin {
+
+	constructor(){
+		this.poller = new Poller({
+			url: 'http://next-navigation.ft.com/v1/taxonomy',
+			refreshInterval: ms('15m')
```
Comment (i-like-robots) [inline src/navigation/hierarchyMixin.js]: Cool module but is it worth it over `1000 * 60 * 15`?

File: main.js [hunk 3]
```diff
@@ -29,6 +29,7 @@ module.exports = function(options) {
 		withFlags: false,
 		withHandlebars: false,
 		withNavigation: false,
+		withNavigationHierarchy: false,
```
Comment (i-like-robots) [inline main.js]: taxonomy?

File: src/navigation/decorate.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+	return false;
+}
+
+function decorateDrawerItems(items, currentUrl){
+	return items.some(item => decorateDrawerItem(item, currentUrl));
+}
+
+function decorateDrawer(items, currentUrl){
+	items.some(section => decorateDrawerItems(section, currentUrl));
```
Comment (i-like-robots) [inline src/navigation/decorate.js]: why `some`?

File: src/navigation/decorate.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
+"use strict";
+
+function isCurrentLink(item, currentUrl){
+	return (item.href && item.href === currentUrl) || (item.id && currentUrl.includes(item.id));
+}
+
+function decorateDrawerItem(item, currentUrl){
```
Comment (i-like-robots) [inline src/navigation/decorate.js]: I don't these methods are specific to the drawer - they should be able to traverse any link structure

File: src/navigation/decorate.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		item.selected = true;
+		return true;
+	}
+
+	return false;
+}
+
+function decorateItemArray(items, currentUrl){
+	items.some(item => decorateItem(item, currentUrl));
+}
```
Comment (i-like-robots) [inline src/navigation/decorate.js]: These do the same?

File: src/navigation/hierarchyMixin.js (hunk 2 above)
Comment (aintgoin2goa) [inline src/navigation/hierarchyMixin.js]: I wouldn't use it client-side but on the server why now?  It brings in no other dependencies and makes the code a bit easier to understand

File: main.js (hunk 3 above)
Comment (aintgoin2goa) [inline main.js]: Yeah, spent some time wondering about this (and googled a definition of taxonomy), hierarchy seems to make more sense here?

File: src/navigation/decorate.js (hunk 4 above)
Comment (aintgoin2goa) [inline src/navigation/decorate.js]: So that, once we find the current link we can stop looking

File: src/navigation/decorate.js (hunk 5 above)
Comment (aintgoin2goa) [inline src/navigation/decorate.js]: Yeah, the drawer has a different structure to the others (nested arrays, headings and `item.item`) was easier to just split it out

File: src/navigation/decorate.js (hunk 6 above)
Comment (aintgoin2goa) [inline src/navigation/decorate.js]: Sorry?

File: src/navigation/hierarchyMixin.js (hunk 2 above)
Comment (i-like-robots) [inline src/navigation/hierarchyMixin.js]: just cold feet over left-pad-gate

**Pull request merged**